### PR TITLE
Add initial code

### PR DESCRIPTION
**Pull request by ricardogama**
This PR adds the initial code of the plugin:
- Plugin souce code
- Test suite
- Documentation
- Support files
  - package.json
  - .gitignore
  - .eslintrc.yml
  - .babelrc
  - .travis.yml

### CODE REVIEW

**Comment by ricardogama** [comment]
@joaogranado @abelsoares Please take a look.

File: .babelrc [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "presets": ["es2015", "stage-2"],
+  "plugins": [
```
Comment (nunorafaelrocha) [inline .babelrc]: Don't need to break.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  ],
+  "options": {
+    "isparta": "--report html --dir test/coverage",
+    "mocha": "--compilers js:babel-register --bail --require should test/index.js"
+  },
+  "scripts": {
+    "build": "rm -rf dist && babel src --out-dir dist",
+    "changelog": "github_changelog_generator --bug-labels --enhancement-labels --header-label='# Changelog'",
+    "cover": "NODE_ENV=test babel-node node_modules/.bin/isparta cover $npm_package_options_isparta _mocha -- $npm_package_options_mocha",
+    "eslint": "git diff --cached --name-only --diff-filter=ACMRTUXB | grep -E '\\.(js)(\\..+)?$' | xargs eslint",
```
Comment (nunorafaelrocha) [inline package.json]: I would add `jscs` as well.

File: package.json [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    "test": "NODE_ENV=test mocha $npm_package_options_mocha"
```
Comment (nunorafaelrocha) [inline package.json]: `"lint": "eslint src test && jscs src test"`

File: package.json [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    "pre-commit": "^1.1.2",
+    "should": "^8.3.1",
+    "sinon": "^1.17.3"
+  },
+  "engines": {
+    "iojs": ">= 1.0.0",
+    "node": ">= 0.10"
+  },
+  "pre-commit": {
+    "run": [
```
Comment (nunorafaelrocha) [inline package.json]:
``` js
"pre-commit": [
  "lint"
]
```

File: test/index.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    await clearTables();
+  });
+
+  after(async () => {
+    await dropTables();
+  });
+
+  it('should throw an error if model has no registered dependents', async () => {
+    const author = await repository.Model.extend({ tableName: 'Author' }).forge().save();
+    await Post.forge().save({ authorId: author.get('id') });
```
Comment (nunorafaelrocha) [inline test/index.js]: New line.

File: test/index.js [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+      should.fail();
+    } catch(e) {
+      e.constraint.should.equal('post_authorid_foreign');
+    }
+  });
+
+  it('should throw an error if model has dependents and `cascadeDelete` option is given as `false`', async () => {
+    const author = await Author.forge().save();
+    await Post.forge().save({ authorId: author.get('id') });
```
Comment (nunorafaelrocha) [inline test/index.js]: Same as above.

File: test/index.js [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    } catch(e) {
+      e.constraint.should.equal('post_authorid_foreign');
+    }
+  });
+
+  it('should not delete model and its dependents if an error is thrown on destroy', async () => {
+    const author = await Author.forge().save();
+    const post1 = await Post.forge().save({ authorId: author.get('id') });
+    const post2 = await Post.forge().save({ authorId: author.get('id') });
+    await Comment.forge().save({ postId: post1.get('id') });
```
Comment (nunorafaelrocha) [inline test/index.js]: Same as above. Perhaps something to check on the whole PR?

File: test/index.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+  it('should throw an error if model has no registered dependents', async () => {
+    const author = await repository.Model.extend({ tableName: 'Author' }).forge().save();
+    await Post.forge().save({ authorId: author.get('id') });
+
+    try {
+      await author.destroy();
+
+      should.fail();
+    } catch(e) {
```
Comment (nunorafaelrocha) [inline test/index.js]: Space after `catch`.

File: test/index.js [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+  it('should throw an error if model has dependents and `cascadeDelete` option is given as `false`', async () => {
+    const author = await Author.forge().save();
+    await Post.forge().save({ authorId: author.get('id') });
+
+    try {
+      await author.destroy({ cascadeDelete: false });
+
+      should.fail();
+    } catch(e) {
```
Comment (nunorafaelrocha) [inline test/index.js]: Same as above.

File: test/index.js [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    await Comment.forge().save({ postId: post1.get('id') });
+    await Comment.forge().save({ postId: post2.get('id') });
+
+    sinon.stub(Model, 'destroy').throws(new Error('foobar'));
+
+    try {
+      await author.destroy();
+
+      should.fail();
+    } catch(e) {
```
Comment (nunorafaelrocha) [inline test/index.js]: Same as above.

File: test/index.js [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    sinon.stub(Model, 'destroy').throws(new Error('foobar'));
+
+    try {
+      await repository.knex.transaction(transaction => Author.forge().save(null, { transacting: transaction })
+        .then(() => Author.forge().save(null, { transacting: transaction }))
+        .then(author => author.destroy({ transacting: transaction }))
+      );
+
+      should.fail();
+    } catch(e) {
```
Comment (nunorafaelrocha) [inline test/index.js]: Same as above.

File: test/utils/index.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Export `recreateTables`.
+ */
+
+export function recreateTables() {
+  return repository.knex.schema
+    .dropTableIfExists('Account')
+    .dropTableIfExists('Comment')
+    .dropTableIfExists('Post')
+    .dropTableIfExists('Author')
```
Comment (nunorafaelrocha) [inline test/utils/index.js]: Sort?

File: test/utils/index.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * Export `clearTables`.
+ */
+
+export async function clearTables() {
+  await repository.knex('Account').del();
+  await repository.knex('Comment').del();
+  await repository.knex('Post').del();
+  await repository.knex('Author').del();
```
Comment (nunorafaelrocha) [inline test/utils/index.js]: Sort?

File: test/utils/index.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Export `dropTables`.
+ */
+
+export function dropTables() {
+  return repository.knex.schema
+    .dropTable('Account')
+    .dropTable('Comment')
+    .dropTable('Post')
+    .dropTable('Author');
```
Comment (nunorafaelrocha) [inline test/utils/index.js]: Sort?

File: README.md [hunk 15]
```diff
@@ -1 +1,87 @@
 # bookshelf-cascade-delete
+
+This [Bookshelf.js](https://github.com/tgriesser/bookshelf) plugin brings you cascade delete with a simple configuration on your models.
```
Comment (joaogranado) [inline README.md]: ... plugin provides cascade delete...

File: README.md [hunk 16]
```diff
@@ -1 +1,87 @@
...
+```sh
+$ npm install --save bookshelf-cascade-delete
+```
+
+## Usage
+
+Require and register the `bookshelf-cascade-delete` plugin:
+
+```js
+var bookshelf = require('bookshelf')(knex);
```
Comment (joaogranado) [inline README.md:22]: And what about providing an example with `import` as well?

File: README.md [hunk 17]
```diff
@@ -1 +1,87 @@
...
+var cascadeDelete = require('bookshelf-cascade-delete');
+
+// You need to access the `default` property since the plugin is transpilled from es6 modules syntax.
+bookshelf.plugin(cascadeDelete.default);
+```
+
+Define which relations depend on your model when it's destroyed with the `dependents` prototype property:
+
+```js
+var Post = bookshelf.Model.extend({
```
Comment (joaogranado) [inline README.md]: `const`.

File: README.md [hunk 18]
```diff
@@ -1 +1,87 @@
...
+  tableName: 'Post'
+});
+
+var Author = bookshelf.Model.extend({
```
Comment (joaogranado) [inline README.md]: Same as above.

File: README.md [hunk 19]
```diff
@@ -1 +1,87 @@
...
+  tableName: 'Author',
+  posts: function() {
+    return this.hasMany(Post);
+  }
+}, {
+  dependents: ['posts']
+});
+```
+
+**NOTE:** This plugin extends the `destroy` method of Bookshelf's `Model`, so if you are also extending or overriding it on your models make sure to call its prototype after your work is done:
```
Comment (joaogranado) [inline README.md]: remove `also`.

File: src/index.js [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+ * Export `bookshelf-cascade-delete` plugin.
+ */
+
+export default Bookshelf => {
+  const Model = Bookshelf.Model.prototype;
+
+  Bookshelf.Model = Bookshelf.Model.extend({
+    cascadeDelete(transaction, options) {
+      return Promise.map(this.constructor.recursiveDeletes(this.get('id'), options), query => query(transaction))
+        .then(() => Model.destroy.call(this, defaults({
```
Comment (joaogranado) [inline src/index.js]: WDYT about using spread properties instead of `_.defaults`?

File: src/index.js [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+            key: foreignKey,
+            model: target
+          }
+        };
+      }, {});
+    },
+    recursiveDeletes(parent, options) {
+      // Stringify in case of parent being an instance of query.
+      const parentValue = typeof parent === 'number' || typeof parent === 'string' ? `'${parent}'` : parent.toString();
+      const queries = reduce(this.dependencyMap(), (result, dependent) => {
```
Comment (joaogranado) [inline src/index.js]: Could you provide a description for this line?

File: README.md (hunk 16 above)
Comment (ricardogama) [inline README.md:22]: I think that doesn't bring much value to the documentation, and would make it too extensive.

Comment (joaogranado) [inline README.md:22]: LGTM.

**Pull request merged**